wallet/src/de/schildbach/wallet/service/WalletTransactionMetadataProvider.kt

```diff
@@ -53,82 +59,150 @@ class WalletTransactionMetadataProvider @Inject constructor(
                 TransactionCategory.fromTransaction(type, txValue, isInternal),
-                taxCategory = null,
-                exchangeRate?.fiat?.currencyCode,
-                exchangeRate?.fiat?.let {
-                    MonetaryFormat.FIAT.noCode().format(it).toString()
-                }
+                taxCategory = platformTaxCategory?.let { TaxCategory.fromValue(it) },
+                currencyCode = code,
+                rate = rate,
+                memo = myMemo,
+                service = platformService
             )
             transactionMetadataDao.insert(metadata)
+            // only add to the change cache if some metadata exists
+            if (metadata.isNotEmpty() && !isSyncingPlatform && hasChanges) {
+                transactionMetadataChangeCacheDao.insert(TransactionMetadataCacheItem(metadata))
+            }
             log.info("txmetadata: inserting $metadata")
         }
     }
 
-    private suspend fun updateAndInsertIfNotExist(txId: Sha256Hash, update: suspend () -> Unit) {
+    private suspend fun updateAndInsertIfNotExist(txId: Sha256Hash, isSyncingPlatform: Boolean, update: suspend () -> Unit) {
         if (transactionMetadataDao.exists(txId)) {
+            log.info("txmetadata for $txId exists, only do update")
             update()
         } else {
-            insertTransactionMetadata(txId)
+            log.info("txmetadata for $txId does not exist, perform insert, then update")
+            insertTransactionMetadata(txId, isSyncingPlatform)
             update()
         }
     }
 
     override suspend fun importTransactionMetadata(txId: Sha256Hash) {
-        updateAndInsertIfNotExist(txId) { }
+        updateAndInsertIfNotExist(txId, false) { }
     }
 
     override suspend fun setTransactionMetadata(transactionMetadata: TransactionMetadata) {
         transactionMetadataDao.insert(transactionMetadata)
     }
 
-    override suspend fun setTransactionTaxCategory(txId: Sha256Hash, taxCategory: TaxCategory) {
-        updateAndInsertIfNotExist(txId) {
+    override suspend fun setTransactionTaxCategory(txId: Sha256Hash, taxCategory: TaxCategory, isSyncingPlatform: Boolean) {
+        updateAndInsertIfNotExist(txId, isSyncingPlatform) {
             transactionMetadataDao.updateTaxCategory(txId, taxCategory)
+            if (!isSyncingPlatform) {
+                transactionMetadataChangeCacheDao.insertTaxCategory(txId, taxCategory)
+            }
         }
     }
 
-    override suspend fun setTransactionType(txId: Sha256Hash, type: Int) {
+    override suspend fun setTransactionSentTime(
+        txId: Sha256Hash,
+        timestamp: Long,
+        isSyncingPlatform: Boolean
+    ) {
+        updateAndInsertIfNotExist(txId, isSyncingPlatform) {
+            transactionMetadataDao.updateSentTime(txId, timestamp)
+            if (!isSyncingPlatform) {
+                transactionMetadataChangeCacheDao.insertSentTime(txId, timestamp)
+            }
+        }
+    }
```

Comment:
`isSyncingPlatform` is true when this method is called from `PlatformSyncService` -- this will prevent an infinite loop of sending new data to platform which is then read, possibly triggering those same changes to be sent again.